src/Workspaces/Core/Portable/Execution/AbstractReferenceSerializationService.cs

```diff
@@ -231,13 +231,18 @@ public AnalyzerReference ReadAnalyzerReferenceFrom(ObjectReader reader, Cancella
             throw ExceptionUtilities.UnexpectedValue(type);
         }
 
-        private void WriteAnalyzerFileReferenceMvid(AnalyzerFileReference reference, ObjectWriter writer, CancellationToken cancellationToken)
+        private void WriteAnalyzerFileReferenceMvid(
+            AnalyzerFileReference file, ObjectWriter writer, bool usePathFromAssembly, CancellationToken cancellationToken)
         {
             try
             {
-                // use actual assembly path rather than one returned from reference.FullPath
+                cancellationToken.ThrowIfCancellationRequested();
+
+                // use actual assembly path rather than one returned from reference.FullPath if asked (usePathFromAssembly)
                 // 2 can be different if analyzer loader used for the reference do something like shadow copying
-                var assemblyPath = TryGetAnalyzerAssemblyPath(reference) ?? reference.FullPath;
+                // otherwise, use reference.FullPath. we use usePathFromAssembly == false for vsix installed analyzer dlls
+                // to make sure we don't load them up front and they don't get shadow copied.
```

Comment:
> ... to make sure we don't load them up front ...

❓ This code loads an assembly regardless of the value specified in `usePathFromAssembly`. How can the change impact the set of assemblies which get loaded?

Reply:
why it load dlls regardless of what is used in usePathFromAssembly? I had same issue before which is why usePathFromAssembly and WorkspaceAnalyzerReference was added at the first place. why you believe it will load the assembly?

Reply:
> So the problem seems to be the call to `TryGetAnalyzerAssemblyPath`

This was not obvious to me. It looks like `TryGetAnalyzerAssemblyPath` would just be a quick method (maybe a `Path.Combine` or related, but nothing super expensive), followed by the use of a `FileStream` and `PEReader` that look like the source of "loading the assembly". It would be good to update the comment to more clearly explain the value of `usePathFromAssembly`, especially considering some users won't have access to RPS runs.

Reply:
Sure